inspirehep/modules/relations/__init__.py

```diff
@@ -0,0 +1,78 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of INSPIRE.
+# Copyright (C) 2015, 2016 CERN.
+#
+# INSPIRE is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# INSPIRE is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with INSPIRE; if not, write to the Free Software Foundation, Inc.,
+# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+
+"""Relations."""
+
+from flask import current_app
+
+from werkzeug.urls import url_parse
+from invenio_db import db
+from invenio_records.signals import after_record_insert, after_record_update
+
+from .models import Relations
+
+def uri_to_uuid(uri):
+    parsed_uri = url_parse(uri)
+
+    path_parts = parsed_uri.path.strip('/').split('/')
+    if len(path_parts) < 2:
+        current_app.logger.error('Bad JSONref URI: {0}'.format(uri))
+        return None
+
+    record_type = path_parts[-2]
+    recid = path_parts[-1]
+    return PersistentIdentifier.get(record_type, recid).id
+
+
+def crawl_json_for_ref(json, root='', relation=''):
```

Comment:
Recursively crawl a json looking for all the `$ref`s and keeping track of the corresponding JSON path (e.g. `authors[0]`) and relation (e.g. `authors`).